util/topsql/stmtstats/aggregator.go

```diff
@@ -0,0 +1,162 @@
+
+// StatementStatsRecord is the merged StatementStatsMap with timestamp.
+type StatementStatsRecord struct {
+	Timestamp int64
+	Data      StatementStatsMap
+}
+
+// aggregator is used to collect and aggregate data from all StatementStats.
+// It is responsible for collecting data from all StatementStats, aggregating
+// them together, uploading them and regularly cleaning up the closed StatementStats.
+type aggregator struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	statsSet   sync.Map // map[*StatementStats]struct{}
+	collectors sync.Map // map[Collector]struct{}
+	running    int32
+}
+
+// newAggregator creates an empty aggregator.
+func newAggregator() *aggregator {
+	return &aggregator{}
+}
+
+// run will block the current goroutine and execute the main loop of aggregator.
+func (m *aggregator) run() {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	atomic.StoreInt32(&m.running, 1)
+	defer func() {
+		atomic.StoreInt32(&m.running, 0)
+	}()
+	tick := time.NewTicker(time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+		case <-tick.C:
+			m.aggregate()
+		}
+	}
+}
+
+// aggregate data from all associated StatementStats.
+// If StatementStats has been closed, collect will remove it from the map.
+func (m *aggregator) aggregate() {
+	r := StatementStatsRecord{
+		Timestamp: time.Now().Unix(),
+		Data:      StatementStatsMap{},
+	}
+	m.statsSet.Range(func(statsR, _ interface{}) bool {
+		stats := statsR.(*StatementStats)
+		if stats.Finished() {
+			m.unregister(stats)
+		}
+		r.Data.Merge(stats.Take())
+		return true
+	})
+	m.collectors.Range(func(c, _ interface{}) bool {
+		c.(Collector).CollectStmtStatsRecords([]StatementStatsRecord{r})
+		return true
+	})
+}
+
+// register binds StatementStats to aggregator.
+// register is thread-safe.
+func (m *aggregator) register(stats *StatementStats) {
+	m.statsSet.Store(stats, struct{}{})
+}
+
+// unregister removes StatementStats from aggregator.
+// unregister is thread-safe.
+func (m *aggregator) unregister(stats *StatementStats) {
+	m.statsSet.Delete(stats)
+}
+
+// registerCollector binds a Collector to aggregator.
+func (m *aggregator) registerCollector(collector Collector) {
+	m.collectors.Store(collector, struct{}{})
+}
+
+// unregisterCollector removes Collector from aggregator.
+func (m *aggregator) unregisterCollector(collector Collector) {
+	m.collectors.Delete(collector)
+}
+
+// close ends the execution of the current aggregator.
+func (m *aggregator) close() {
+	m.cancel()
+}
+
+// closed returns whether the aggregator has been closed.
+func (m *aggregator) closed() bool {
+	return atomic.LoadInt32(&m.running) == 0
+}
+
+// SetupAggregator is used to initialize the background aggregator goroutine of the stmtstats module.
+func SetupAggregator() {
+	if v := globalAggregator.Load(); v != nil {
+		go v.(*aggregator).run()
+		return
+	}
+	c := newAggregator()
+	go c.run()
+	globalAggregator.Store(c)
+}
```

Comment:
If global is unavoidable, then how about:

```go
var globalAggregator  = ...

func SetupAggregator() {
    globalAggregator.Start()
}

func CloseAggregator() {
    globalAggregator.Stop()
}

func (m *aggregator) Start() {
    started = m.started.compareAndSwap(...)
    if !started {
        go m.runLoop()
    }
}
```

In this way you can avoid a lot of places checking `globalAggregator.Load(); v != nil`.

Reply:
It is now feasible, due to the removal of the extra logic in `newAggregator()`.
done for this.